File: tyama_codeiq711_iter.py
#!/usr/bin/python
#coding:utf-8

def find(cond,a):
	for e in a:
		if cond(e): return e
	return None
def permutation(x,n=None):
	if n==None: n=len(x)
	if n<0 or len(x)<n: return
	a=sorted(x)
	yield tuple(a[0:n])
	while True:
		a=a[0:n]+a[len(a)-1:n-1:-1]
		k=find(lambda i: a[i]<a[i+1], range(len(a)-2,-1,-1))
		if k==None: break
		l=find(lambda i: a[k]<a[i], range(len(a)-1,k,-1))
		a[k],a[l]=a[l],a[k]
		a=a[0:k+1]+a[len(a)-1:k:-1]
		yield tuple(a[0:n])

N=6
e0=[0]*N+[1]*N
f0=[0]*N+[1]*N
#各Pは経路を表す。1が縦、0が横を表す。
r=0
for _e in permutation(e0):
	for _f in permutation(f0):
		flg=0
		zero1=0
		zero2=N
		one1=0
		one2=N
		for i in range(N*2):
			zero1 += 1 if _e[i]==0 else 0
			one1  += 1 if _e[i]==1 else 0
			zero2 -= 1 if _f[N*2-1-i]==0 else 0
			one2  -= 1 if _f[N*2-1-i]==1 else 0
			if zero1==zero2: flg+=1
			if one1==one2: flg+=1
		if flg>=2: r+=1
print(r) # 527552

File: tyama_codeiq711_iter.pl
#!/usr/bin/perl
use strict;
sub permute(&@){
	my $code = shift;
	my @a = @_;
	while($code->(@a)){
		my $i;
		#push(@a,reverse splice @a, $n);
		for($i=$#a-1;$i>=0;$i--){if($a[$i]<$a[$i+1]){last;}}
		if($i<0){return;}
		my $k=$i;
		for($i=$#a;$i>=$k+1;$i--){if($a[$k]<$a[$i]){last;}}
		my $l=$i;
		@a[$k,$l]=@a[$l,$k];
		push(@a, reverse splice @a, $k+1);
	}
}

my $N=6;
my @e0;
my @f0;
my $i;
my $r=0;
for($i=0;$i<$N;$i++){$e0[$i]=$f0[$i]=0;}
for($i=0;$i<$N;$i++){$e0[$N+$i]=$f0[$N+$i]=1;}
permute {
	my @_e=@_;
	permute {
		my $flg=0;
		my $zero1=0;
		my $zero2=$N;
		my $one1=0;
		my $one2=$N;
		for($i=0;$i<$N*2;$i++){
			if($_e[$i]==0){$zero1++;}
			if($_e[$i]==1){$one1++;}
			if($_[$N*2-1-$i]==0){$zero2--;}
			if($_[$N*2-1-$i]==1){$one2--;}
			if($zero1==$zero2){$flg++;}
			if($one1==$one2){$flg++;}
		}
		if($flg>=2){$r++;}
		1;
	} @f0;
	1;
} @e0;
print $r."\n";